### DISCUSSION TITLE
Dealing with lots of columns: Folders of columns?

### DISCUSSION THREAD

**Discussion by cool-RR** [saulpw/visidata#1688] [Q&A]
I want to brainstorm with you about a problem and potential solutions.

I'm writing experiments that produce time series data. These are ML experiments in which a bunch of AI agents are interacting with each other in some environment, and I collect metrics about their behavior. I decide what kind of metrics to collect, i.e. which columns to add to the CSV. Sometimes I have 5 columns, sometimes I have 30 columns.

I'm often tempted to collect more and more metrics, so I could possibly analyze them later. If I keep more data, I can make more interesting analyses later. Some metrics might bring me to 100 columns.

The issue is that a CSV with 100 columns is difficult to explore in VisiData, and in most of its competitors. While I have the ability to hide columns, it'll still be frustrating to have to hide a bunch of columns every time I open a CSV file.

One thought that I had is to create folders of columns. Instead of seeing all the columns by default, it'll just show the folders of columns, so instead of seeing 100 columns, you might see 5 column folders. Then you could expand each one, and they might have subfolders as well. Getting the UI right for such a functionality would be challenging. 

Do you have any thought about this problem, proposed solution or any other solutions?

Thanks,
Ram.

**Comment by saulpw** [comment]
Do you have control over the output format or header names?

If you can output in JSON, you could make the "folders of columns" as dicts, and then use `(` and `)` to expand and contract these column groups.  This will currently work.

If you can change the column names, one idea I've had is to collapse columns with a common subprefix (e.g `foo.a` and `foo.b`) into one unexpanded `foo` column group that would act just like the above. This would require some changes in the csv loader but wouldn't be too big a deal.

**Reply by cool-RR** [reply]
Off-topic: I just discovered that in all our discussions here, I accidentally posted a new answer instead of a comment. I feel pretty stupid :laughing: 

> Do you have control over the output format or header names?

Yes I do.

> If you can output in JSON, you could make the "folders of columns" as dicts, and then use `(` and `)` to expand and contract these column groups. This will currently work.

This is great! Thank you. 

> If you can change the column names, one idea I've had is to collapse columns with a common subprefix (e.g `foo.a` and `foo.b`) into one unexpanded `foo` column group that would act just like the above. This would require some changes in the csv loader but wouldn't be too big a deal.

Yes I would like that! Should I open a feature request?

**Reply by cool-RR** [reply]
For future reference: I opened this feature request for it: https://github.com/saulpw/visidata/issues/1702
